src/sema/mod.rs

```diff
@@ -386,6 +408,121 @@ fn resolve_pragma(
+
+        let doccomments = parse_doccomments(comments, doc_comment_start, loc.start());
+
+        if let pt::SourceUnitPart::ContractDefinition(contract) = part {
+            let mut parts = Vec::new();
+            let mut parts_annotations: Vec<&pt::Annotation> = Vec::new();
+            let mut doc_comment_start = contract.loc.start();
+
+            for part in &contract.parts {
+                match part {
+                    pt::ContractPart::Annotation(note) => parts_annotations.push(note),
+                    _ => {
+                        let tags =
+                            parse_doccomments(comments, doc_comment_start, part.loc().start());
+
+                        parts.push(ContractPart {
+                            part,
+                            doccomments: tags,
+                            annotations: parts_annotations,
+                        });
+                        parts_annotations = Vec::new();
+                    }
+                }
+
+                if let pt::ContractPart::FunctionDefinition(f) = &part {
+                    if let Some(pt::Statement::Block { loc, .. }) = &f.body {
+                        doc_comment_start = loc.end();
+                        continue;
+                    }
+                }
+
+                doc_comment_start = part.loc().end();
+            }
+
+            if !parts_annotations.is_empty() {
+                for note in parts_annotations {
+                    ns.diagnostics.push(ast::Diagnostic::error(
+                        note.loc,
+                        "annotations should precede 'constructor' or other item".into(),
+                    ));
+                }
+            }
+
+            contracts.push(ContractDefinition {
+                contract_no,
+                loc: contract.loc,
+                ty: contract.ty.clone(),
+                annotations,
+                doccomments,
+                name: contract.name.as_ref(),
+                base: contract.base.clone(),
+                parts,
+            });
+
+            contract_no += 1;
+        } else {
+            items.push(SourceUnitPart {
+                doccomments,
+                annotations,
+                part,
+            });
+        }
+        annotations = Vec::new();
+
+        if let pt::SourceUnitPart::FunctionDefinition(f) = part {
+            if let Some(pt::Statement::Block { loc, .. }) = &f.body {
+                doc_comment_start = loc.end();
+                continue;
+            }
+        }
+
+        doc_comment_start = part.loc().end();
+    }
+
+    if !annotations.is_empty() {
+        for note in annotations {
+            ns.diagnostics.push(ast::Diagnostic::error(
+                note.loc,
+                "annotations should precede 'contract' or other item".into(),
+            ));
+        }
+    }
+
+    SourceUnit { items, contracts }
+}
+
+fn no_annotations(annotations: &[&pt::Annotation], item: &str, ns: &mut ast::Namespace) {
```

Comment:
I'd prefer a more descriptive name (alternatively a comment describing what the function does). What about renaming this to `disallow_annotations`?

Reply:
Good point, I've renamed it to `fn annotions_not_allowed()`